storage/google/cloud/storage/blob.py

```diff
@@ -1951,6 +1982,84 @@ def updated(self):
             return _rfc3339_to_datetime(value)
 
 
+class StreamingFile(FileIO):
+    """File-like object, that provides streaming data.
+
+    When read() is called, it checks if the requested part of the
+    file has already been downloaded. If not, it downloads chunk
+    large enough to satisfy the read() call and then reads itself.
+
+    filename (str): The name of the file to which data will be downloaded.
+
+    download_url (str): The URL where the media can be accessed.
+
+    headers (dict): (Optional) Headers to be sent with the request(s).
+
+    transport (:class:`~google.auth.transport.requests.AuthorizedSession`):
+        The transport (with credentials) that will
+        make authenticated requests.
+
+    chunk_size (int): The size of chunks to be downloaded.
+    """
+    def __init__(self, filename, download_url, headers, transport, chunk_size):
+        super(StreamingFile, self).__init__(filename, 'wb+')
+
+        self._transport = transport
+        self._download = ChunkedDownload(
+            download_url,
+            chunk_size,
+            self,
+            headers=headers
+        )
+
+    def read(self, size=None):
+        """Overriding the base class read() method to organize streaming.
+
+        The function checks whether the part that the user is trying to read has already
+        been downloaded. If not, it calculates the missing part
+        then downloads and redirects it to super().read().
+
+        Args:
+            size (int): (Optional) Size of the block to be read.
+
+        Returns:
+            Bytes read from the file.
+        """
+        if not self._download.finished:
+            cursor_place = self.tell()
+
+            if size is None:  # download whole file
+                self.seek(*_EOF)
+                while not self._download.finished:
+                    self._download_chunk()
+            else:  # download requested part if needed
+                # calculating the missing part
+                can_be_read = os.fstat(self.fileno()).st_size - cursor_place
+                to_download = size - can_be_read
+
+                if to_download > 0:
+                    chunk_size = self._download.chunk_size
+                    self._download.chunk_size = to_download
+                    self.seek(*_EOF)
+                    self._download_chunk()
+                    self._download.chunk_size = chunk_size
+
+            self.seek(cursor_place, 0)
+
+        return super(StreamingFile, self).read(size)
```

Comment:
The main idea is to download the part, that user is trying to read, but which is not yet downloaded. Before download we should remember cursors position, then we must move cursor to the end of file (to avoid writing downloaded data to the middle of the file in case if user made `seek()`). And then we should brought cursor back to users position and call `super().read()`